### ISSUE TITLE
Allow local config to inherit from other config

### ISSUE THREAD

**Issue by slarse** [repobee/repobee#874]
In order to make it less of a hassle to manage multiple courses that have some configuration in common, it would be nice to be able to inherit options from a parent config. There are three aspects to this: how is inheritance defined, what to do about transitive inheritance, and whether or not to allow shadowing.

Finally, there is also the question of how the `config` commands should work.

Assume that the improved config lookup of #873 is already implemented for the further discussion.

Also assume this directory structure:

```
A
├── B
│   ├── C
│   │   └── repobee.ini
│   └── repobee.ini
└── repobee.ini
```

Where `A`, `B`, and `C` are directories, each containing a config file. Saying e.g. "config A" refers to the config file at A/repobee.ini, and config B is at A/B/repobee.ini, etc.

### How to define inheritance
The first choice here is whether inheritance should be implicit or explicit. We could say that config C implicitly inherits from config B, which in turn implicitly inherits from config A. **I think that's grounds for potentially massive confusion.**

With explicit inheritance, an inheriting config file would have to specify some option. We can have completely nominal inheritance, such that config C can inherit from config A for example by specifying an option `parent = ../../repobee.ini`. We could also have the same structural inheritance as with the implicit option outlined above, except that the config would need to specify e.g. `inherit = True`.

I'm thinking that nominal inheritance is most flexible and leaves least risk of disaster. It's a very explicit choice. On the other hand, explicit structural inheritance sounds kind of neat.

### Is inheritance transitive?
If config C inherits from config B, and config B inherits from config A, does then config C also inherit from config A? If this is _not_ the case, is it config C then even allowed to inherit from config B, or does that cause an error?

I think we should either have transitive inheritance, or inheriting from a config that also inherits should cause an error.

### What about shadowing?
If config B defines `some_option = value`, and it inherits from config A that defines `some_option = other_value`, what happens? I'm thinking we have two options.

1. Error: multiple values for option `some_option`
2. The value in config B shadows the one in config A

While I think option 2 _could_ be useful, it could also cause confusion. Option 1 leaves very little room for error.

### How should the config commands work?
* I think `config show` should be enhanced to show which config file each configured value comes from.
* How will `config wizard` work?
    - One option would be to allow the user to choose which config file to edit, when editing a config file that inherits from another.
    - Another option would be that it simply works as-is: it edits the most local config file by default, and to edit another

**Comment by slarse** [comment]
@dmusican @htdebeer @rjglasse I've discussed this briefly with all of you. @tohanss you might also want to chime in.

I'm decided on _explicit_ inheritance, but not whether to go nominal (`parent = path`) or structural (`inherit = True`). I'm leaning toward nominal. Preferences?

I'm very much undecided on transitive inheritance. On the one hand, it's easy to implement, possibly easier than disallowing even. On the other hand, I don't see much of a use for it. Also, imagine forgetting that config C inherits from B, and then have B inherit from A, which suddenly also affects config C would be a possible cause for disaster.

I'm similarly undecided on shadowing. I can see how it could be useful, and semantically it would not be unlike how command line options currently shadow config options. Perhaps shadowing could also be opt-in (e.g. `allow_shadowing = True`)?

Any thoughts and ideas are very much welcome!

**Comment by htdebeer** [comment]
The nominal variant of inheritance is more flexible, and would allow inheritance not only in a parent-child configuration, but in any configuration imaginable, including child-parent! I.e., that the parent directory inherits from a child directory's repobee.ini. In a more realistic example, I could have a directory with "base" repobee.ini files, like "production.ini" and "test.ini", with URLs, tokens, and user names for my production environment and my test environment, and then I can just swap these out in "child" configuration depending on if I want to target test or production.

I do think transitive inheritance can make sense, but if we can have programmatically access to another RepoBee configuration, we can copy all the relevant fields to the child configuration. If that's a route we like to take, then shadowing does not make much sense either. If a field is specified in this configuration use it, otherwise look for it in the parent.

[This](https://github.com/htdebeer-tue/repobee-canvas/issues/2) feature request for repobee-canvas will likely make use of this inheritance functionality. As I envision it now, each Canvas course will have a directory with a RepoBee configuration. Each assignment in the course will have a subdirectory with a RepoBee configuration that inherits from the course configuration.

**Comment by slarse** [comment]
> The nominal variant of inheritance is more flexible, [...]

That's a pretty convincing argument for nominal inheritance! With that in mind, I think nominal is the clear-cut winner.

> I do think transitive inheritance can make sense, but if we can have programmatically access to another RepoBee configuration, we can copy all the relevant fields to the child configuration.

While programmatic access is entirely possible (we bake it into the as-of-yet undocumented `Config` class), this has to be designed with a non-Python-programmer user in mind. It should be accessible to any RepoBee user. I.e. programmatic access shouldn't play a role in deciding how inheritance for config files works.

> If that's a route we like to take, then shadowing does not make much sense either. If a field is specified in this configuration use it, otherwise look for it in the parent

This is precisely what I mean by shadowing. The option/value combination in the child config shadows that in the parent config (or _overrides_, could also be the term). The alternative (no shadowing allowed), would be for RepoBee to scream bloody murder when it finds multiple. The former is more flexible, the latter leaves less room for error. The middle road would be to make shadowing opt-in, such that an exception is raised on finding the same option/value in child and parent _unless_ the child defines e.g. `allow_shadowing = True`.

> [This](https://github.com/htdebeer-tue/repobee-canvas/issues/2) feature request for repobee-canvas will likely make use of this inheritance functionality. As I envision it now, each Canvas course will have a directory with a RepoBee configuration. Each assignment in the course will have a subdirectory with a RepoBee configuration that inherits from the course configuration.

That looks like an excellent use case to act as a driver for implementing all of this.

**Comment by dmusican** [comment]
I'm late to this party and am just weighing in by invitation, but I agree that nominal inheritance makes sense. In addition to the above use cases, I can easily imagine having a template directory with basic settings, and then sibling directories that want to inherent from it. Nominal inheritance would even easily allow for multiple inheritance... which hasn't been addressed yet?

I'm struggling a little with the transitive question because I don't have a current personal use case for it.  But for what it's worth, with nominal inheritance, my mental model would expect transitive dependence to be there. Git itself has an "include" an option in its own config files, which just pretend to insert the config options from the other file. Being consistent with that would seem to have some value. You could try to be explicit about what's going on when viewing the config, at least. A user who wanted transitive inheritance would be frustrated when repobee would whimsically declare "I can't do that."

Regarding shadowing: I think again it would be most intuitive to me if it mimicked Git's behavior, which would be (I think?) to just shadow without saying anything. I think the justification is pretty much the same as for transitive inheritance; if someone wants it, it would seem awfully pedantic to prevent it. ("Repobee knows what's best.") And again, showing that in a view of the config could help.

I don't have any skin in any of this currently, but I could easily imagine ending up there next year with two classes using repobee simultaneously. But even with that, I'm not seeing myself in the near future using any more than two levels (one level of inheritance), so for me at the moment this particular issue is fairly academic. If the Canvas integration has a need for it, I completely agree that it should drive it.

**Comment by slarse** [comment]
@dmusican 

> Nominal inheritance would even easily allow for multiple inheritance... which hasn't been addressed yet?

I hadn't even considered allowing multiple inheritance. I'd veto that, though, it adds significant complexity at little to no benefit.

> I'm struggling a little with the transitive question because I don't have a current personal use case for it. But for what it's worth, with nominal inheritance, my mental model would expect transitive dependence to be there.

I'm also leaning towards this simply because that's the "reasonable" thing to expect. If it's not too much of a hassle to implement, I suppose I'll do it. It should be easy; recursive config resolution coupled with a cyclic inheritance check.

 
> Regarding shadowing: I think again it would be most intuitive to me if it mimicked Git's behavior, which would be (I think?) to just shadow without saying anything.

Yes, I can personally see use cases for this as well. For example, say you have a course with some basic settings and a students list of all students. Then you want to have subdirectories where you work with "groups" of students (perhaps you have a couple of exercise groups), then it'd make sense to be able to override the students list.
 
Basically, I'm going to give the following a try:

* Inheritance form: Nominal, but not multiple
* Transitive? Yes.
* Shadowing? Yes.
* Config category command extensions
    - Show which options belong in which config file when running `config show`.
    - `config wizard` will for the time being only edit the "most local" config file, for the sake of simplicity.
